QC Hospitalisations by age

mschoettle opened the pull request:
Adds weekly hospitalisations by age group (per 100k).
Hovering also shows the absolute number of new hospitalisations for that week.

<img width="674" alt="Screen Shot 2021-12-10 at 06 40 04" src="https://user-images.githubusercontent.com/9328433/145568810-ac36243b-da4c-4a38-8ddf-ce94fa4b04a1.png">

mschoettle committed:
Adds data processing function to extract QC data by age group.

mschoettle committed:
Adds weekly new hospitalisations by age group for QC.

Places it next to MTL cases by age group figure.
Limited to the last 16 weeks in order to not overcrowd the figure.
Data is normalized to 100k per age group.

jeremymoreau committed:
Add range sliders to QC hospitalisations by age and MTL new cases distribution by age figures

jeremymoreau commented:
This looks great, thanks! Added range sliders like for the other figures (instead of limiting data to last 16 weeks):
![screenshot1](https://user-images.githubusercontent.com/7446564/146902874-9818d729-5fc5-426a-ac06-442e62d5b819.png)
![screenshot2](https://user-images.githubusercontent.com/7446564/146902899-4ed2e12f-194c-4f2d-819c-816d623d0fb9.png)


Issues remaining:
1. y-axis doesn't autoscale when adjusting range. There doesn't seem to be an way to do this. Manual zooming with the drag-to-zoom tool still works.
2. Seems like the latest week of data isn't displayed anymore after adding the range slider.
3. Might be nice to set a default range for the range slider (other than all) once there's a lot of data? https://github.com/plotly/plotly.py/issues/828#issuecomment-329005535
4. Is there a reason to display the week count? This seems arbitrary since it's just the number of weeks since data started being collected (and redundant with the date). I commented it out for now, but let me know if there is a reason I'm missing.

mschoettle commented:
Thanks!

What I originally wanted to do is to start with an initial range of the last 16 weeks but because it does not zoom automatically the y-axis was too large and I ended up limiting it to the 16 weeks.
I don't even think that there is way to do this programmatically, right?

>     2. Seems like the latest week of data isn't displayed anymore after adding the range slider.

Is this for both figures? I think I removed the last 2 days of the hospitalization data because it is delayed and likely underreported.

>     4. Is there a reason to display the week count? This seems arbitrary since it's just the number of weeks since data started being collected (and redundant with the date). I commented it out for now, but let me know if there is a reason I'm missing.

The week number is the week number within the year. I added it because the data is weekly to show which week the data point corresponds to. But it could also just say "Week of yyyy-mm-dd".

mschoettle commented:
>     1. y-axis doesn't autoscale when adjusting range. There doesn't seem to be an way to do this. Manual zooming with the drag-to-zoom tool still works.

I tried `figure.update_layout(xaxis=dict(range=['2021-11-01', '2021-12-08']), yaxis=dict(range=[0, 20]))` to test but unfortunately, contrary to the x-axis, the y-axis stays at the defined range.

Another problem with the range sliders is that there are added some additional empty dates at the beginning and end.
 
> 2. Seems like the latest week of data isn't displayed anymore after adding the range slider.

Removing `core.py` line 304 (`data_qc_age = data_qc_age.iloc[:-2]`) might resolve this.

jeremymoreau committed:
Fix rangeslider including empty dates

Manually sets the min/max date range for the mtl_age_fig and qc_age_fig to the earliest/latest dates in the dataset.

jeremymoreau commented:
> What I originally wanted to do is to start with an initial range of the last 16 weeks but because it does not zoom automatically the y-axis was too large and I ended up limiting it to the 16 weeks.
> I don't even think that there is way to do this programmatically, right?

> I tried `figure.update_layout(xaxis=dict(range=['2021-11-01', '2021-12-08']), yaxis=dict(range=[0, 20]))` to test but unfortunately, contrary to the x-axis, the y-axis stays at the defined range.

I tried searching for a solution, but it seems that it is not yet implemented in Plotly:
https://github.com/plotly/plotly.js/issues/6958
https://community.plotly.com/t/autoscaling-the-y-axis-when-using-rangeslider/1456

Also some discussion for plotly.js (but not solved yet):
https://github.com/plotly/plotly.js/issues/5544
https://github.com/plotly/plotly.js/issues/1876

One possible solution could be to not use the Plotly rangeslider and instead use a Dash slider component outside the figure with a callback to redraw the Plotly figure, e.g. https://dash.plotly.com/basic-callbacks#dash-app-layout-with-figure-and-slider
Not the simplest solution though... so maybe we leave it as is for now with the manual zoom? Ideally Plotly would add the option, but I don't know if/when they will add it.


> The week number is the week number within the year. I added it because the data is weekly to show which week the data point corresponds to. But it could also just say "Week of yyyy-mm-dd".

That makes sense! I removed it initially because the week count labels overlaped with the date labels on the x axis (when the number of date labels increases above a certain number). I think adding "Week of yyyy-mm-dd" in the hover template title is a good option. The title of the plots also says the data is weekly.

> Is this for both figures? I think I removed the last 2 days of the hospitalization data because it is delayed and likely underreported.
> Removing core.py line 304 (data_qc_age = data_qc_age.iloc[:-2]) might resolve this.

Yes, that was it! Removing the last 2 days hides the last week of data (so currently the MTL plot goes until Dec 6 while the QC hospitalisations fig goes to Nov 29 if the last 2 days are removed). Not sure which is best, commented it out for now so that both figures end on the same date.

> Another problem with the range sliders is that there are added some additional empty dates at the beginning and end.

Not sure why those empty dates are being added (used the same function as all the other figures with a range slider), but manually setting the x axis range seems to resolve this.
